src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -864,27 +866,62 @@ export class MatrixRTCSession extends TypedEventEmitter<MatrixRTCSessionEvent, M
+        const stateKey = legacy ? localUserId : this.makeMembershipStateKey(localUserId, localDeviceId);
         try {
-            await this.client.sendStateEvent(
-                this.room.roomId,
-                EventType.GroupCallMemberPrefix,
-                newContent,
-                legacy ? localUserId : this.makeMembershipStateKey(localUserId, localDeviceId),
-            );
+            await this.client.sendStateEvent(this.room.roomId, EventType.GroupCallMemberPrefix, newContent, stateKey);
             logger.info(`Sent updated call member event.`);
 
             // check periodically to see if we need to refresh our member event
-            if (this.isJoined() && legacy) {
-                this.memberEventTimeout = setTimeout(this.triggerCallMembershipEventUpdate, MEMBER_EVENT_CHECK_PERIOD);
+            if (this.isJoined()) {
+                if (legacy) {
+                    this.memberEventTimeout = setTimeout(
+                        this.triggerCallMembershipEventUpdate,
+                        MEMBER_EVENT_CHECK_PERIOD,
+                    );
+                } else {
+                    await this.client
+                        ._unstable_sendStateFuture(
+                            this.room.roomId,
+                            {
+                                // TODO: choose a reasonable timeout
+                                future_timeout: 10000,
+                            },
+                            EventType.GroupCallMemberPrefix,
+                            {}, // leave event
+                            stateKey,
+                        )
+                        .then((res) => {
+                            this.cancelFutureToken = res.cancel_token;
+                            this.scheduleRefreshFuture(res.refresh_token);
+                        })
+                        .catch((e) => logger.error("Failed to send future:", e));
+                }
+            } else if (!legacy && this.cancelFutureToken !== undefined) {
+                // TODO: remove this once the Synapse MSC4140 implementation cancels timeout futures on state change
+                await this.client
+                    ._unstable_useFutureToken(this.cancelFutureToken)
+                    .catch((e) => logger.error("Failed to cancel future:", e));
             }
         } catch (e) {
             const resendDelay = CALL_MEMBER_EVENT_RETRY_DELAY_MIN + Math.random() * 2000;
-            logger.warn(`Failed to send call member event: retrying in ${resendDelay}`);
+            logger.warn(`Failed to send call member event (retrying in ${resendDelay}): ${e}`);
             await new Promise((resolve) => setTimeout(resolve, resendDelay));
             await this.triggerCallMembershipEventUpdate();
         }
     }
 
+    private scheduleRefreshFuture(refreshToken: string): void {
+        // TODO: choose a reasonable timeout
+        this.memberEventTimeout = setTimeout(() => this.refreshFuture(refreshToken), 3000);
```

Comment:
So the heartbeat ping is every 3 seconds and the state event timeout 10seconds?
I think 2x should be fine. 5 and 10 seconds or 3 and 6 seconds.

Do you have any "choosing mechanism" in mind?

Reply:
I don't have any particular mechanism in mind, other than not wanting to spam network requests (but to still send enough to be useful :slightly_smiling_face: )

Reply:
then lets do 5 8 as a starter. As an experiment if this already causes calls to be ended  accidentally by bad network because there is only a 3 second buffer between the timeout and the refresh?